pkg/costmodel/metrics.go

```diff
@@ -228,11 +283,323 @@ func (s ServiceMetric) Write(m *dto.Metric) error {
+					if !ok {
+						klog.V(4).Infof("Unable to find parameters for storage class \"%s\". Does pv \"%s\" have a storageClassName?", pv.Spec.StorageClassName, pv.Name)
+					}
+					var region string
+					if r, ok := pv.Labels[v1.LabelZoneRegion]; ok {
+						region = r
+					} else {
+						region = defaultRegion
+					}
+					cacPv := &costAnalyzerCloud.PV{
+						Class:      pv.Spec.StorageClassName,
+						Region:     region,
+						Parameters: parameters,
+					}
+					GetPVCost(cacPv, pv, a.Cloud, region)
+					c, _ := strconv.ParseFloat(cacPv.Cost, 64)
+					a.PersistentVolumePriceRecorder.WithLabelValues(pv.Name, pv.Name).Set(c)
+					labelKey := getKeyFromLabelStrings(pv.Name, pv.Name)
+					pvSeen[labelKey] = true
+				}
+			}
+			for labelString, seen := range nodeSeen {
+				if !seen {
+					klog.Infof("Removing %s from nodes", labelString)
+					labels := getLabelStringsFromKey(labelString)
+					ok := a.NodeTotalPriceRecorder.DeleteLabelValues(labels...)
+					if ok {
+						klog.Infof("removed %s from totalprice", labelString)
+					} else {
+						klog.Infof("FAILURE TO REMOVE %s from totalprice", labelString)
+					}
+					ok = a.CPUPriceRecorder.DeleteLabelValues(labels...)
+					if ok {
+						klog.Infof("removed %s from cpuprice", labelString)
+					} else {
+						klog.Infof("FAILURE TO REMOVE %s from cpuprice", labelString)
+					}
+					ok = a.GPUPriceRecorder.DeleteLabelValues(labels...)
+					if ok {
+						klog.Infof("removed %s from gpuprice", labelString)
+					} else {
+						klog.Infof("FAILURE TO REMOVE %s from gpuprice", labelString)
+					}
+					ok = a.RAMPriceRecorder.DeleteLabelValues(labels...)
+					if ok {
+						klog.Infof("removed %s from ramprice", labelString)
+					} else {
+						klog.Infof("FAILURE TO REMOVE %s from ramprice", labelString)
+					}
+					delete(nodeSeen, labelString)
+				} else {
+					nodeSeen[labelString] = false
+				}
+			}
+			for labelString, seen := range containerSeen {
+				if !seen {
+					labels := getLabelStringsFromKey(labelString)
+					a.RAMAllocationRecorder.DeleteLabelValues(labels...)
+					a.CPUAllocationRecorder.DeleteLabelValues(labels...)
+					a.GPUAllocationRecorder.DeleteLabelValues(labels...)
+					delete(containerSeen, labelString)
+				} else {
+					containerSeen[labelString] = false
+				}
+			}
+			for labelString, seen := range pvSeen {
+				if !seen {
+					labels := getLabelStringsFromKey(labelString)
+					a.PersistentVolumePriceRecorder.DeleteLabelValues(labels...)
+					delete(pvSeen, labelString)
+				} else {
+					pvSeen[labelString] = false
+				}
+			}
+			for labelString, seen := range pvcSeen {
+				if !seen {
+					labels := getLabelStringsFromKey(labelString)
+					a.PVAllocationRecorder.DeleteLabelValues(labels...)
+					delete(pvcSeen, labelString)
+				} else {
+					pvcSeen[labelString] = false
+				}
+			}
+
+			select {
+			case <-time.After(time.Minute):
+			case <-recordingStop:
```

Comment:
Does the `close(recordingStop)` somehow send something on this channel? I'm confused about how the stop mechanism works. This `select` for combo timer/escape mechanism is cool.

Reply:
`close(recordingStop)` will signal anything currently receiving the channel and anything that will receive the channel in the future (ie: it auto resolves). This is especially nice if you you have 1 channel and many consumers and you need all consumer goroutines to "complete" at once.

Reply:
I believe that receiving on a channel actually returns the `T` value as well as `bool` for whether or not the receive was successful or closed. 
```go
    ch := make(chan string)

    go func() {
        result, ok := <- ch
        if ok {
            fmt.Printf("Successful Receive: %s\n", result)
        } else {
            fmt.Println("Channel Closed")
        }
    }()
    time.Sleep(2 * time.Second)
    close(ch)
    time.Sleep(1 * time.Second)
```

I haven't ever used the `_, ok` on receive before but I think this is how `range` over a channel is able to tell when to stop iterating.

Reply:
While I'm on this tangent, I recognized that it would probably be "better" to use a `chan struct{}` but honestly, this grosses me out.